test/lazy/test_ddp.py

```diff
@@ -0,0 +1,149 @@
+# Owner(s): ["oncall: jit"]
+
+import copy
+import os
+import sys
+import torch
+import torch._lazy
+import torch._lazy.ts_backend
+import torch.distributed as dist
+import torch.nn as nn
+import torch.optim as optim
+
+from datetime import timedelta
+from torch._utils_internal import TEST_MASTER_ADDR as MASTER_ADDR
+from torch._utils_internal import TEST_MASTER_PORT as MASTER_PORT
+from torch.nn.parallel import DistributedDataParallel as DDP
+from torch.testing._internal.common_utils import run_tests, TEST_WITH_DEV_DBG_ASAN, NO_MULTIPROCESSING_SPAWN
+from torch.testing._internal.common_distributed import MultiProcessTestCase
+
+if not dist.is_available():
+    print("Distributed not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+if not torch.cuda.is_available() or not dist.is_nccl_available():
+    print("CUDA/NCCL not available, skipping tests", file=sys.stderr)
+    sys.exit(0)
+
+if TEST_WITH_DEV_DBG_ASAN:
+    print("Skip dev-asan as torch + multiprocessing spawn have known issues", file=sys.stderr)
+    sys.exit(0)
+
+if NO_MULTIPROCESSING_SPAWN:
+    print("Spawn not available, skipping tests.", file=sys.stderr)
+    sys.exit(0)
+
+torch._lazy.ts_backend.init()
+os.environ["WORLD_SIZE"] = str(torch.cuda.device_count())
+
+class ToyModel(nn.Module):
+    def __init__(self):
+        super(ToyModel, self).__init__()
+        self.net1 = nn.Linear(10, 10)
+        self.relu = nn.ReLU()
+        self.net2 = nn.Linear(10, 5)
+
+    def forward(self, x):
+        return self.net2(self.relu(self.net1(x)))
+
+
+def step(model, input, labels, device, loss_fn, optimizer):
+    input = input.to(device)
+    labels = labels.to(device)
+    optimizer.zero_grad()
+
+    output = model(input)
+
+    loss = loss_fn(output, labels)
+    loss.backward()
+    optimizer.step()
+    # FIXME(alanwaketan): Investigate why we produce different
+    # results if mark_step.
+    # if device.type == "lazy":
+    #     torch._lazy.mark_step(str(device))
+    return loss
+
+def init(model, device, rank, size):
+    model = copy.deepcopy(model).to(device)
+    if device.type == "lazy":
+        # Somehow, LazyTensor crashes if gradient_as_bucket_view is not set.
+        # FIXME(alanwaketan): Investigate why and if we can remove this constraint.
+        model = DDP(model, gradient_as_bucket_view=True)
```

Comment:
what does using 'gradient_as_bucket_view' actually mean, and can you claim that it is ok/optimal to do so for lazy?

Reply:
According to the [documentation](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel), it should be optimal except for one case where users want to detach the gradients. 
> gradient_as_bucket_view ([bool](https://docs.python.org/3/library/functions.html#bool)) – When set to True, gradients will be views pointing to different offsets of allreduce communication buckets. This can reduce peak memory usage, where the saved memory size will be equal to the total gradients size. Moreover, it avoids the overhead of copying between gradients and allreduce communication buckets. When gradients are views, detach_() cannot be called on the gradients. If hitting such errors, please fix it by referring to the [zero_grad()](https://pytorch.org/docs/stable/generated/torch.optim.Optimizer.zero_grad.html#torch.optim.Optimizer.zero_grad) function in torch/optim/optimizer.py as a solution. Note that gradients will be views after first iteration, so the peak memory saving should be checked after first iteration.

Reply:
hmm. ok, i think it is fine to leave this way for now.